### PR TITLE
UCS/SYS: event_set API waits number of max_events events.

### PR DESCRIPTION
**Pull request by hiroyuki-sato**
## What

extend event_set API for waiting number of specified events.

## Why ?

Please see this discussion
https://github.com/openucx/ucx/pull/3610#discussion_r294858659

For testing #3698, I create separate PR.

Hello, @yosefe and @dmitrygladkov.
Could you please give me some advice on how to fix the test?

Before, merging this PR, we need to modify `uct_tcp_iface_progress` function in `src/uct/tcp/tcp_iface.c`, otherwise the following test will fail like the below.

```
[----------] 6 tests from tcp/test_uct_event_fd
[ RUN      ] tcp/test_uct_event_fd.am/0


common/test_helpers.cc:46: Failure
Failed
Connection timed out - abort testing
make: *** [test] Aborted (core dumped)
```

I modified this function like the below. but this test can't pass yet.
Does `uct_tcp_iface_progress` function sometime don't triger event?

```C
unsigned uct_tcp_iface_progress(uct_iface_h tl_iface)
{
    uct_tcp_iface_t *iface = ucs_derived_of(tl_iface, uct_tcp_iface_t);
    unsigned count         = 0;
    unsigned nevents       = 0;
    ucs_status_t status;

    status = ucs_event_set_wait(iface->event_set,
                                1, /* We need to fix this part */
                                0, uct_tcp_iface_handle_events,
                                (void *)&count, &nevents);
    ucs_trace_poll("iface=%p ucs_event_set_wait(): "
                   "read events=%u, status=%d", iface, nevents, status);
    return count;
}
```

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
ok to test

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10723/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>FAILED</b> on 33 of 33 workers (click for details)</summary>

**Note:** the logs will be deleted after 27-Jun-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/a3fb9dfee6b5429a4ea639c22685102c526d333c/_main) | :x: FAILURE
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/b37c1c55f6e1408311d61dbc52c02791959d1cfc/hpc-arm-cavium-jenkins_W0) | :x: FAILURE
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/04441f5570f3bd5e3be13a2bcafc39b0df45b01e/hpc-arm-cavium-jenkins_W1) | :x: FAILURE
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/14ef0d8fd00a005e0863ea298c4df263bfaa823f/hpc-arm-cavium-jenkins_W2) | :x: FAILURE
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/9b135000540ef40d19366c014ae467eacb5543e8/hpc-arm-cavium-jenkins_W3) | :x: FAILURE
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/4b71cbb5256166801101160fafdf9b807d0f6f1e/hpc-arm-hwi-jenkins_W0) | :x: FAILURE
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/dce3262699d93f5039f782bc9aae5a862560f41a/hpc-arm-hwi-jenkins_W1) | :x: FAILURE
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/0561903828a142ad6a82c1ca392de9de61e50d5c/hpc-arm-hwi-jenkins_W2) | :x: FAILURE
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/1b30281c30161c4504f14fa0a3b9fa170b2c72c5/hpc-arm-hwi-jenkins_W3) | :x: FAILURE
[hpc-test-node-upstream_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/559f4c88efd9a06abe697f19c8e570cef0794e9c/hpc-test-node-upstream_W0) | :x: FAILURE
[hpc-test-node-upstream_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/8c26950aa32c2f28180f4b7cdbeb1859787f9a36/hpc-test-node-upstream_W1) | :x: FAILURE
[hpc-test-node-upstream_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/0f879dce61906f0dc71da922b64503f7e7306f15/hpc-test-node-upstream_W2) | :x: FAILURE
[hpc-test-node-upstream_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/45dfb1f617166ccf26d14bd5399f769217772508/hpc-test-node-upstream_W3) | :x: FAILURE
[hpc-test-node2_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/87c4d0bad1fdf06386069192a86eaff7da02d961/hpc-test-node2_W0) | :x: FAILURE
[hpc-test-node2_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/e59b298c148a926f8e62102db9c055ca12d1dfd5/hpc-test-node2_W1) | :x: FAILURE
[hpc-test-node2_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/064826239c49f9be61633d1f3b832cc66ce95d4e/hpc-test-node2_W2) | :x: FAILURE
[hpc-test-node2_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/1ce0088bfe4b697776aaed0c5d1468b3908bb684/hpc-test-node2_W3) | :x: FAILURE
[hpc-test-node3_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/80ca35a339cbd54059e44987ce9e8ce03446343f/hpc-test-node3_W0) | :x: FAILURE
[hpc-test-node3_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/35233676976d538c116170490d1a00d94de80ff6/hpc-test-node3_W1) | :x: FAILURE
[hpc-test-node3_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/300290d5ec08d01738650435d5075ede665eed7e/hpc-test-node3_W2) | :x: FAILURE
[hpc-test-node3_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/adf298e9022f261bc9e47e262607472fd54310d1/hpc-test-node3_W3) | :x: FAILURE
[hpc-test-node_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/ef06d042d71a9ec933e7eec2e0177150bd2d13c0/hpc-test-node_W0) | :x: FAILURE
[hpc-test-node_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/5c85e4bf3b88a8a047cd02f8dba24dc7a2eaa690/hpc-test-node_W1) | :x: FAILURE
[hpc-test-node_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/97176e195091d732e261d7dab05a064c81a063bd/hpc-test-node_W2) | :x: FAILURE
[hpc-test-node_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/b1473c421c6059aeaee2e4cf1b8939e0c93f353d/hpc-test-node_W3) | :x: FAILURE
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/f01fdb2335d7582331da37b10cce12f0f0fd5b43/r-vmb-ppc-jenkins_W0) | :x: FAILURE
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/9eee8c2ede57baec27856cf3f8dd5fca8159a32f/r-vmb-ppc-jenkins_W1) | :x: FAILURE
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/ae741809da50434813e1d5e9b4d618eb0dd25d03/r-vmb-ppc-jenkins_W2) | :x: FAILURE
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/4d70e3c145043a40293a789c07948ba86d8c7cc2/r-vmb-ppc-jenkins_W3) | :x: FAILURE
[vulcan-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/4e3efef50215f02dd8ce37f0408ba598ac4c0bfd/vulcan-jenkins_W0) | :x: FAILURE
[vulcan-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/04e8e7696603ae1246ac0d30de04e37f997d85d9/vulcan-jenkins_W1) | :x: FAILURE
[vulcan-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/9962bcff2f04dbcea1c693fe2345a7d8a3e534a6/vulcan-jenkins_W2) | :x: FAILURE
[vulcan-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/161f0b3d369b518b7fbecc0d3c2885c0/raw/a8e837757b6ca0182f6df70272ddbe1fcfc222a5/vulcan-jenkins_W3) | :x: FAILURE

File: src/ucs/sys/event_set.c [hunk 1]
```diff
@@ -142,38 +142,36 @@ ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
...
+	for (i = 0; i < nready; i++) {
+	    events = ucs_event_set_map_to_events(ep_events[i].events);
+	    event_set_handler(ep_events[i].data.ptr, events, arg);
+	}
+
+	*read_events    += nready;
+        max_wait_events -= nready;
+        /* After 1st iteration, epoll_wait always returns immediately */
+        timeout_wait_ms = 0;
+    } while (max_wait_events > 0);
```
Comment (dmitrygladkov) [inline src/ucs/sys/event_set.c]: how we protected against the endless loop if we can't read more, but `max_wait_events > 0`
we should think about how to overcome it

Comment (hiroyuki-sato) [reply src/ucs/sys/event_set.c]: It seems that there are two use cases.
* wait until `max_events` events occur.
* wait specific time and allow timeout(no event happen case)

Just idea, `ucs_event_set_wait` user choice which mode use.
For example, introduce a new option `mode`. the second case returns immediately after `timeout_ms` waits.(except EINTR case)

But even the first case, It requires TIMEOUT for avoiding an endless loop.

...Finally, the `mode` option needless.  🤔

File: src/ucs/sys/event_set.c [hunk 2]
```diff
@@ -142,38 +142,36 @@ ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                 void *arg, unsigned *read_events)
 {
     struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
-    unsigned max_wait_events;
+    unsigned max_wait_events = max_events;
     int nready, events, i;
+    int timeout_wait_ms = timeout_ms;
```
[the remark this answers is not in the document]
Comment (hiroyuki-sato) [reply src/ucs/sys/event_set.c]: I used this parameter for implementing @yosefe idea. Do you have any idea to implement it without `timeout_wait_ms`? How to switch the first iteration and others?

https://github.com/openucx/ucx/pull/3610#discussion_r294917067
> maybe ucs_event_set_wait() should pass some timeout in 1st iteration, and pass timeout=0 in all next iterations, to read the remaining events

File: src/ucs/sys/event_set.c [hunk 3]
```diff
@@ -142,38 +142,36 @@ ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
...
-        event_set_handler(ep_events[i].data.ptr, events, arg);
-    }
-
-    *read_events = nready;
-    if  ((nready < max_events) && (nready == UCS_EVENT_EPOLL_MAX_EVENTS)) {
-        return UCS_INPROGRESS;
-    }
+    *read_events = 0;
+    do {
+	nready = epoll_wait(event_set->epfd, ep_events, max_wait_events,
```
[the remark this answers is not in the document]
Comment (hiroyuki-sato) [reply src/ucs/sys/event_set.c]: Thanks, Fixed on the local. I'll push it later.

**Comment by hiroyuki-sato** [comment]
Hello, @dmitrygladkov May I close this PR? This PR refactor on #3746

**Comment by dmitrygladkov** [comment]
> Hello, @dmitrygladkov May I close this PR? This PR refactor on #3746

Hello, @hiroyuki-sato.
Yes, pls close the PR as #3746 replaces it

**Comment by hiroyuki-sato** [comment]
Thank you!, I close this PR.

**Pull request closed without merging**